src/Shared/FileUtilitiesRegex.cs

```diff
@@ -34,5 +34,180 @@ internal static class FileUtilitiesRegex
+            }
+
+            bool prevCharWasSlash = true;
+            bool hasSubfolder = false;
+
+            for (int i = 2; i < pattern.Length; i++)
+            {
+                if (pattern[i] == Path.DirectorySeparatorChar ||
+                    pattern[i] == Path.AltDirectorySeparatorChar)
+                {
+                    if (prevCharWasSlash)
+                    {
+                        //We get here in the case of an extra slash.
+                        return false;
+                    }
+
+                    hasSubfolder = true;
+                    prevCharWasSlash = true;
+                }
+                else
+                {
+                    if (hasSubfolder)
+                    {
+                        //A character after a subfolder confirms the beginning of a unc pattern
+                        return true;
+                    }
+
+                    prevCharWasSlash = false;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Indicates whether the file-spec begins with a UNC pattern and how long the match is. -1 indicates no match.
+        /// </summary>
+        /// <param name="pattern"></param>
+        /// <returns></returns>
+        internal static int DoesStartWithUncPatternMatchLength(string pattern)
+        {
+            if (!MeetsUncPatternMinimumRequirements(pattern))
+            {
+                return -1;
+            }
+
+            bool prevCharWasSlash = true;
+            bool hasSubfolder = false;
+
+            for (int i = 2; i < pattern.Length; i++)
+            {
+                if (pattern[i] == Path.DirectorySeparatorChar ||
+                    pattern[i] == Path.AltDirectorySeparatorChar)
+                {
+                    if (prevCharWasSlash)
+                    {
+                        //We get here in the case of an extra slash.
+                        return -1;
+                    }
+                    else if(hasSubfolder)
+                    {
+                        return i;
+                    }
+
+                    hasSubfolder = true;
+                    prevCharWasSlash = true;
+                }
+                else
+                {
+                    prevCharWasSlash = false;
+                }
+            }
+
+            if(!hasSubfolder)
+            {
+                //no subfolder means no unc pattern. string is something like "\\abc" in this case
+                return -1;
+            }
+
+            return pattern.Length;
+        }
+
+        /// <summary>
+        /// Indicates whether or not the file-spec meets the minimum requirements of a UNC pattern.
+        /// UNC pattern requires a minimum length of 5 and first two characters must be a slash.
+        /// </summary>
+        /// <param name="pattern"></param>
+        /// <returns></returns>
+        internal static bool MeetsUncPatternMinimumRequirements(string pattern)
```

Comment:
we can probably be justified sprinkling a few of these around, but definitely here:

```suggestion
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        internal static bool MeetsUncPatternMinimumRequirements(string pattern)
```